apps/aecore/src/aec_miner.erl

```diff
@@ -0,0 +1,230 @@
+%% state name. Whenever a gen_statem receives an event sent using
+%% gen_statem:send_event/2, the instance of this function with the same
+%% name as the current state name StateName is called to handle
+%% the event. It is also called if a timeout occurs.
+%%
+%% @spec state_name(Type, Event, State) ->
+%%                   {next_state, NextStateName, NextState} |
+%%                   {next_state, NextStateName, NextState, [Actions]}
+%% @end
+%%--------------------------------------------------------------------
+idle({call, From}, start, State) ->
+    lager:info("Mining resumed by user.", []),
+    gen_statem:cast(?SERVER, mine),
+    {next_state, running, State, [{reply, From, ok}]};
+idle({call, From}, suspend, State) ->
+    {next_state, idle, State, [{reply, From, {error, not_started}}]};
+idle({call, From}, _Msg, State) ->
+    {next_state, idle, State, [{reply, From, {error, not_supported}}]};
+idle(_Type, _Msg, State) ->
+    {next_state, idle, State}.
+
+running({call, From}, start, State) ->
+    {next_state, running, State, [{reply, From, {error, already_started}}]};
+running({call, From}, suspend, State) ->
+    lager:info("Mining suspended by user.", []),
+    {next_state, idle, State, [{reply, From, ok}]};
+running(cast, mine, State) ->
+    case aec_mining:mine(?MINING_ATTEPTS_PER_CYCLE) of
+        {ok, Block} ->
+            Header = aec_blocks:to_header(Block),
+            case aec_chain:insert_header(Header) of
+                ok ->
+                    case aec_chain:write_block(Block) of
+                        ok ->
+                            lager:info("Reached height ~p.", [Block#block.height]);
+                        {error, Reason} ->
+                            lager:error("Block insertion failed: ~p.", [Reason])
+                    end;
+                {error, Reason} ->
+                    lager:error("Header insertion failed: ~p.", [Reason])
+            end,
+            gen_statem:cast(?SERVER, mine),
+            {next_state, running, State};
+        {error, generation_count_exhausted} ->
+            %% Needs more attempts, go on trying
+            lager:info("Failed to mine block in ~p attampts, retrying.",
+                       [?MINING_ATTEPTS_PER_CYCLE]),
+            gen_statem:cast(?SERVER, mine),
+            {next_state, running, State};
+        {error, key_not_found} ->
+            report_suspended(),
+            gen_statem:cast(?SERVER, check_keys),
+            {next_state, waiting_for_keys, State};
+        {error, Reason} ->
+            lager:error("Mining attempt failed with error: ~p.", [Reason]),
+            gen_statem:cast(?SERVER, mine),
+            {next_state, running, State}
+    end;
+running({call, From}, _Msg, State) ->
+    {next_state, running, State, [{reply, From, {error, not_supported}}]};
+running(_Type, _Msg, State) ->
+    {next_state, idle, State}.
+
+waiting_for_keys(cast, check_keys, State) ->
```

Comment:
do we want to wait for keys or simply shut down the whole app?

Reply:
@sennui  Because of deferred initialization in aec_keys, we cannot just try once and shut down all.

Reply:
@Luca pointed out that keys may be loaded later than the app starts. So I prefer waiting in order to avoid sync problems.